---
title: "basic_usage"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{basic_usage}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(plinkr)
```

In this vignette, I run the 
commands at [http://zzz.bwh.harvard.edu/plink/data.shtml#plink](http://zzz.bwh.harvard.edu/plink/data.shtml#plink) using plinkr.

This vignette will give output only if PLINK is installed:

```{r}
if (!is_plink_installed()) {
  message("PLINK is not installed")
  message("Tip: use 'plinkr::install_plink()'")
}
```

Here is the first example on the PLINK website:

```
plink --file mydata --pheno pheno.raw --assoc --maf 0.05 --out run1
```

The `--file mydata` means that PLINK will look for the files named
`mydata.ped` and `mydata.map`. In this example, we will use these two from
the PLINK v1.7 example files:

```{r}
if (is_plink_installed()) {
  # Check both files exists
  ped_filename <- get_plink_example_filename(
    example_filename = "test.ped",
    plink_version = "1.7"
  )
  map_filename <- get_plink_example_filename(
    example_filename = "test.map",
    plink_version = "1.7"
  )
  # Get the path without extension, can use either ped or map file
  testthat::expect_equal(
    tools::file_path_sans_ext(ped_filename),
    tools::file_path_sans_ext(map_filename)
  )
  # Get the path without extension
  mydata <- tools::file_path_sans_ext(ped_filename)
}
```


The `--pheno pheno.raw` means that PLINK will look for a files named `pheno.raw`. PLINK does not supply this example file, hence we will use one supplied by plinkr:

```{r}
phenotype_filename <- get_plinkr_filename("pheno.raw")
```

The `--out run1` means that PLINK will create a file that starts with  `run1`. In this context, where we test for an association for a quantitive trait, this file will be called `run1.qassoc`. 

```{r}
output_filename_base <- file.path(tempfile(), "run1")
output_filename_base <- "run1"
output_filename_qassoc <- paste0(output_filename_base, ".qassoc")
```

becomes

```{r}
if (is_plink_installed()) {
  plinkr::run_plink(
    args = c(
      "--file", mydata, # mydata.ped and mydata.map
      "--pheno", phenotype_filename,
      "--assoc",
      "--maf", "0.05",
      "--out", output_filename_base
    )
  )
}
```
Now, let's see what actually happened, except for calling PLINK, by showing what was in the input files, as well as what is in the output file:

The `.ped` file:

```{r}
if (is_plink_installed()) {
  knitr::kable(
    read_plink_ped_file(ped_filename)
  )
}
```

Note that the `.ped` file contains a column called `case_control_code`.
In this example, this column will be overruled by our the phenotypic
values we supply.

Now we take a look at the `.map` file,
which maps the single-nucleotide variations (SNVs)
to their position on the DNA:

```{r}
if (is_plink_installed()) {
  knitr::kable(
    read_plink_map_file(map_filename)
  )
}
```

These are:

 * `r names(read_plink_map_file(map_filename))[1]`: 
    the chromosome code or contig name
 * `r names(read_plink_map_file(map_filename))[2]`: 
    Variant identifier
 * `r names(read_plink_map_file(map_filename))[3]`: 
    Position in morgans or centimorgans.
    This value is optional. Zeroes denote it is unused
 * `r names(read_plink_map_file(map_filename))[4]`: 
    Base-pair coordinat

```{r}
if (is_plink_installed()) {
  knitr::kable(
    read_plink_phenotype_file(phenotype_filename)
  )
}
```

Reading the output:

```{r}
if (is_plink_installed()) {
  knitr::kable(
    read_plink_qassoc_file(output_filename_qassoc)
  )
}
```

 * `trait_name`: name of the quantitive trait,
 * `CHR`: Chromosome number
 * `SNP`: SNP identifier
 * `BP`: Physical position (base-pair)
 * `NMISS`: Number of non-missing genotypes
 * `BETA`: Regression coefficient
 * `SE`: Standard error
 * `R2`: Regression r-squared
 * `T`: Wald test (based on t-distribtion)
 * `P`: Wald test asymptotic p-value

## Is the phenotype file mandatory?

No, you can remove it from the command to run PLINK:

```{r}
if (is_plink_installed()) {
  plinkr::run_plink(
    args = c(
      "--file", mydata, # mydata.ped and mydata.map
      "--assoc",
      "--maf", "0.05",
      "--out", output_filename_base
    )
  )
}
```

Instead, PLINK will use the phenotypic trait in the `.map`
file, for the column named `r names(read_plink_map_file(map_filename))[6]`.
These values (always a 1 or a 2), however, are not phenotypic 
values, these are the notation for case-control associations.

As we run PLINK with a case-control study, 
the input is now in `run1.assoc` instead:

```{r}
if (is_plink_installed()) {
  output_filename_assoc <- paste0(output_filename_base, ".assoc")
  knitr::kable(
    read_plink_assoc_file(output_filename_assoc)
  )
}
```

 * `trait_name`: name of the quantitive trait,
 * `CHR`: Chromosome number
 * `SNP`: SNP identifier
 * `BP`: Physical position (base-pair)
 * `NMISS`: Number of non-missing genotypes
 * `BETA`: Regression coefficient
 * `SE`: Standard error
 * `R2`: Regression r-squared
 * `T`: Wald test (based on t-distribtion)
 * `P`: Wald test asymptotic p-value


## Future

Ru-run if (`rerun_plink`) is fun for the future:

```
plink --rerun run1.log --maf 0.1
plink --rerun run1.log --maf 0.1 --out run2
```

Run PLINK from script, not now (`run_plink_from_script`):

```
plink --script myscript1.txt
```

> where the file myscript1.txt is a plain text file containing

```
--ped ..\data\version1\50K\allsamples.ped 
--map ..\data\allmapfiles\finalversion\autosomal.map 
--out ..\results\working\sample-missingness-v1.22
--from rs66537222
--to rs8837323 
--geno 0.25
--maf 0.02
--missing
```

