---
title: "Detect associations with a simulated quantitative trait"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Detect associations with a simulated quantitative trait}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Introduction

In this vignette,
we are going to detect associations between genetic information
and one or more quantitative traits.

We will be using simulated data,
with a simple relationship between genotype and phenotype.
To see how this same analysis is done with more
messy data, see the 'test_assoc_qt' vignette.

## Setup

Set the RNG seed:

```{r set_rng_seed}
set.seed(314)
```


First load `plinkr`:

```{r setup}
library(plinkr)
```

This vignette will build whether or not PLINK is installed

```{r is_plink_installed}
if (is_plink_installed()) {
  message("PLINK is installed")
} else {
  message(
    "PLINK is not installed",
    "",
    "Tip: run 'plinkr::install_plinks()' to do so"
  )
}
```

## Detecting associations for quantitative traits

To do an association (both for a single and multiple traits), 
we need some parameters:

 * The `.map` table or mapping table, which contains the location
   of the single-nucleotide polymorphisms (SNPs)
 * The `.ped` table or pedigree table, which contains the pedigree
   of the individuals and their SNP values
 * The phenotype table, which contains the one or more phenotypes
   of the individuals
 * The minor allele frequency (MAF) threshold for an allele
   to be included in the analysis

Here, we get simple set of parameters, as to be used in testing,
or -in this case- a simple demonstration:
 
```{r}
assoc_qt_params <- create_demo_assoc_qt_data(
  n_individuals = 16
)
```

### The mapping table

```{r}
knitr::kable(assoc_qt_params$data$map_table)
```

### The pedigree table

The PLINK example \code{.ped} contains:

 * the pedigree of the individuals
 * the SNP calls of the individuals

Show only the pedigree:

```{r}
knitr::kable(assoc_qt_params$data$ped_table[, 1:6])
```

Note that the pedigree table has a column called `case_control_code`,
which will be completely ignored.

Show only the SNP calls:

```{r}
knitr::kable(assoc_qt_params$data$ped_table[, -(1:6)])
```


### The phenotype table

The pedigree table has a column called `case_control_code`,
which will be completely ignored, as it only allows to put in
one phenotype. Instead, we use a table of one or more phenotypic
values:

```{r}
knitr::kable(assoc_qt_params$phe_table)
```

The phenotypes are named after their relationship to the
genotype:

 * ´random´: there is no relationship, values are random
 * ´additive´: there is a perfect additive relationship, 
   with a starting value of 10.0,
   where each 'T' increases the trait by 0.5
 * ´epistatic´: there is an epistatic phenotype, 
   with a starting value of 20.0,
   that can only be 21.0 if all four nucleotides on four
   SNPs are the uncommon alleles

### The minor allele frequency (MAF)


```{r}
message(assoc_qt_params$maf)
```

The `maf` (named after the PLINK `--maf` flag)
denotes the minimal frequency an allele must have to be
included in the analysis. 
We keep this at the lowest non-zero value, so that all
SNPs are analysed.

### Detecting the association

With the mapping, pedigree and phenotype table, 
we can detect the association between genotype and the two traits:

```{r}
if (is_plink_installed()) {
  t <- assoc_qt(
    assoc_qt_params = assoc_qt_params
  )
  knitr::kable(t)
}
```

 * `trait_name`: name of the quantitive trait,
 * `CHR`: Chromosome number
 * `SNP`: SNP identifier
 * `BP`: Physical position (base-pair)
 * `NMISS`: Number of non-missing genotypes
 * `BETA`: Regression coefficient
 * `SE`: Standard error
 * `R2`: Regression r-squared
 * `T`: Wald test (based on t-distribution)
 * `P`: Wald test asymptotic p-value

In the fourth row, the association between `snp_2`
and the `additive` phenotype is discovered
correctly, as the `R2` (the R-squared value, i.e. the
proportion of the phenotype that can be explained by the
genotype) gives the highest possible value of 1.0,
which denotes the phenotype can be perfectly explained
by the genotype.


## Cleanup

```{r clear_plinkr_cache_at_end}
clear_plinkr_cache()
```

```{r check_empty_plinkr_folder_at_end}
check_empty_plinkr_folder()
```
