---
title: "Detect associations with a binary trait"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Detect associations with a binary trait}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Introduction

In this vignette,
we are going to detect associations between genetic information
and one binary traits.

We will be using the data from PLINK files, which
is small, albeit a bit messy, where 'messy' is defined
as that there is not a ideal 1-on-1 relation between genotype
and phenotype. 

## Setup

First load `plinkr`:

```{r setup}
library(plinkr)
```

This vignette will build whether or not PLINK is installed

```{r is_plink_installed}
if (is_plink_installed()) {
  message("PLINK is installed")
} else {
  message(
    "PLINK is not installed",
    "",
    "Tip: run 'plinkr::install_plinks()' to do so"
  )
}
```

## Detecting associations for one binary trait

To do an associationfor one binary trait, 
we need some parameters:

 * The `.map` table or mapping table, which contains the location
   of the single-nucleotide polymorphisms (SNPs)
 * The `.ped` table or pedigree table, which contains the pedigree
   of the individuals, their SNP values and their binary
   trait values
 * The minor allele frequence (MAF)

Here, we get simple set of parameters, as to be used in testing,
or -in this case- a simple demonstration:
 
```{r}
assoc_params <- create_test_assoc_params()
```

### The mapping table

```{r}
knitr::kable(assoc_params$map_table)
```

### The pedigree table

The PLINK example \code{.ped} contains, among others, the pedigree of the individuals:

```{r}
knitr::kable(assoc_params$ped_table)
```

The pedigree table has a column called `case_control_code`,
which contains our binary trait.

### The minor allele frequency (MAF)

```{r}
message(assoc_params$maf)
```

The minor allele frequency (MAF) denotes that alleles that
have an occurrence below this MAF will be excluded from
the analysis. As we want to analyse all data, the MAF is
set to the lowest non-zero value.

### Detecting the association

With the mapping, pedigree and phenotype table, 
we can detect the association between genotype and the single trait:

```{r}
if (is_plink_installed()) {
  t <- assoc(
    assoc_params = assoc_params
  )
  knitr::kable(t)
}
```

 * `CHR`: Chromosome number
 * `SNP`: SNP identifier
 * `BP`: Physical position (base-pair)
 * `A1`: Minor allele name (based on whole sample)
 * `F_A`: Frequency of this allele in cases
 * `F_U`: Frequency of this allele in controls
 * `A2`: Major allele name
 * `CHISQ`: Basic allelic test chi-square (1df)
 * `P`: Asymptotic p-value for this test
 * `OR`: Estimated odds ratio (for A1, i.e. A2 is reference)
