---
title: "Detect associations with a quantitative trait"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Detect associations with a quantitative trait}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Introduction

In this vignette,
we are going to detect associations between genetic information
and one or more quantitative traits.

We will be using the data from PLINK files, which
is small, albeit a bit messy, where 'messy' is defined
as that there is not a ideal 1-on-1 relation between genotype
and phenotype. See the 'demo_assoc_qt' vignette
where the same analysis is done with simulated
ideal data.

## Setup

First load `plinkr`:

```{r setup}
library(plinkr)
```

This vignette will build whether or not PLINK is installed

```{r is_plink_installed}
if (is_plink_installed()) {
  message("PLINK is installed")
} else {
  message(
    "PLINK is not installed",
    "",
    "Tip: run 'plinkr::install_plinks()' to do so"
  )
}
```

## Detecting associations for a single trait

To do an association (both for a single and multiple traits), 
we need some parameters:

 * The `.map` table or mapping table, which contains the location
   of the single-nucleotide polymorphisms (SNPs)
 * The `.ped` table or pedigree table, which contains the pedigree
   of the individuals and their SNP values
 * The phenotype table, which contains the one or more phenotypes
   of the individuals
 * The minor allele frequence (MAF)

Here, we get simple set of parameters, as to be used in testing,
or -in this case- a simple demonstration:
 
```{r}
assoc_qt_params <- create_test_assoc_qt_params()
```

### The mapping table

```{r}
knitr::kable(assoc_qt_params$map_table)
```

### The pedigree table

The PLINK example \code{.ped} contains, among others, the pedigree of the individuals:

```{r}
knitr::kable(assoc_qt_params$ped_table)
```

Note that the pedigree table has a column called `case_control_code`,
which will be completely ignored.

### The phenotype table

The pedigree table has a column called `case_control_code`,
which will be completely ignored, as it only allows to put in
one phenotype. Instead, we use a table of one or more phenotypic
values:

```{r}
knitr::kable(assoc_qt_params$phenotype_table)
```

These trait values, that is the values in the column `phenotype_value`,
are clearly artificial. As the name `phenotype_value` we will
rename it to `trait_x`:

```{r}
names(assoc_qt_params$phenotype_table) <-
  c(
    names(assoc_qt_params$phenotype_table)[-3],
    "trait_x"
  )
```

### The minor allele frequency (MAF)

The minor allele frequency (MAF) is set to 5% as a default:

```{r}
message(assoc_qt_params$maf)
```

We leave it as it is.

### Detecting the association

With the mapping, pedigree and phenotype table, 
we can detect the association between genotype and the single trait:

```{r}
if (is_plink_installed()) {
  t <- assoc_qt(
    assoc_qt_params = assoc_qt_params
  )
  knitr::kable(t)
}
```

 * `trait_name`: name of the quantitive trait,
 * `CHR`: Chromosome number
 * `SNP`: SNP identifier
 * `BP`: Physical position (base-pair)
 * `NMISS`: Number of non-missing genotypes
 * `BETA`: Regression coefficient
 * `SE`: Standard error
 * `R2`: Regression r-squared
 * `T`: Wald test (based on t-distribution)
 * `P`: Wald test asymptotic p-value

## Detecting associations for multiple traits

To detect associations for multiple traits,
we can simply run multiple associations for single traits.

Within `plinkr`, however, 
all we need to do is to extent the phenotype table,
then re-run the analysis.

### The phenotype table

We simply add another column with random values:
 
```{r}
if (is_plink_installed()) {
  assoc_qt_params$phenotype_table$trait_y <-
    seq_len(nrow(assoc_qt_params$phenotype_table)) *
    seq_len(nrow(assoc_qt_params$phenotype_table))
  knitr::kable(assoc_qt_params$phenotype_table)
}
```

### Detecting the association

With the mapping, pedigree and phenotype table, 
we can detect the association between genotype and the two traits:

```{r}
if (is_plink_installed()) {
  t <- assoc_qt(
    assoc_qt_params = assoc_qt_params
  )
  knitr::kable(t)
}
```

 * `trait_name`: name of the quantitive trait,
 * `CHR`: Chromosome number
 * `SNP`: SNP identifier
 * `BP`: Physical position (base-pair)
 * `NMISS`: Number of non-missing genotypes
 * `BETA`: Regression coefficient
 * `SE`: Standard error
 * `R2`: Regression r-squared
 * `T`: Wald test (based on t-distribution)
 * `P`: Wald test asymptotic p-value
