---
title: "use_cases"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{use_cases}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(plinkr)
```



```
for bim in {101..196}
do   
echo  $bim
 for panel in "ONC2_" "NEU_" "CVD2_" "CVD3_" "INF_"
  do
     for chr in  {1..22}
     do
	 awk '$7>=0.01 && $1<30 && $12<5E-08 ' output/$panel$bim.$chr.conditional.assoc.txt.assoc.txt > Sign/Sign.cond.$panel$bim.$chr

	 if [ -s Sign/Sign.cond.$panel$bim.$chr ]
	 then
	  #   echo "File not empty"
	     awk '{if(min>$12||NR==1){min=$12;a[$12]=$0}}END{print a[min]}' Sign/Sign.cond.$panel$bim.$chr > Sign/Top.Sign.cond.$panel$bim.$chr
	     rm Sign/Sign.cond.$panel$bim.$chr
	 else
	   #  echo "File empty"
	     rm  Sign/Sign.cond.$panel$bim.$chr
	 fi


     done 

done
done


cat Sign/Top.Sign* > Top.Sign.All
PLINKin="/my_folder/"

awk '{print $2}' Top.Sign.All > Top.Sign.All.snps

for chr in {1..22}
do
plink --bfile $PLINKin$chr/$chr --extract Top.Sign.All.snps  --recode 12 --out CovarFiles/$chr
done
```

Or:

```
plink --bfile /my_folder/dataBIALLELIC/WGS-SNP-biallelic --clump /my_folder/GENABEL/PEA3/FILTERED/$bio.filtered.pea3.genabel --clump-p1 1e-8 --clump-r2 0.05 --clump-kb 2000 --out /my_folder/GENABEL/PEA3/$bio.r0.05; done
plink --bfile /my_folder/dataBIALLELIC/WGS-SNP-biallelic --clump /my_folder/GENABEL/PEA3/FILTERED/$bio.filtered.pea3.genabel --clump-p1 1e-8 --clump-r2 0.1 --clump-kb 2000 --out /my_folder/GENABEL/PEA3/$bio.r0.1; done
plink --bfile /my_folder/dataBIALLELIC/WGS-SNP-biallelic --clump /my_folder/GENABEL/PEA3/FILTERED/$bio.filtered.pea3.genabel --clump-p1 1e-8 --clump-r2 0.1 --clump-kb 15000 --out /my_folder/GENABEL/PEA3/$bio.15Mb; done
plink --vcf /my_folder/Recal_SNP_INDEL_clean_samples_RENAMED.vcf --make-bed --keep-allele-order --autosome --set-missing-var-ids @:#[b37] --biallelic-only list --out /my_folder/WGS-SNP-INDEL-biallelic
plink --bfile /my_folder/WGS-SNP-biallelic --freq --hardy --allow-extra-chr --out /my_folder/WGS-SNP-biallelic
plink --bfile /my_folder/WGS-SNP-biallelic --autosome --recode transpose --out /my_folder/WGS-SNP-biallelic
plink --bfile /my_folder/dataBIALLELIC/WGS-SNP-biallelic --chr 1-6 --recode transpose --out GenABEL-biallelic-CHR1-6
plink --bfile /my_folder/dataBIALLELIC/WGS-SNP-biallelic --chr 7-14 --recode transpose --out GenABEL-biallelic-CHR7-14
plink --bfile /my_folder/dataBIALLELIC/WGS-SNP-biallelic --chr 15-12 --recode transpose --out GenABEL-biallelic-CHR15-22
plink --bfile /my_folder/WGS-SNP-biallelic --allow-extra-chr --chr 1-22 --extract subsetSNP --recode transpose --out /my_folder/WGS-SNP-biallelic-subset
plink --bfile /my_folder/dataBIALLELIC/WGS-SNP-biallelic --indep-pairwise 10000kb 1 0.8 --maf 0.0015 --out /my_folder/WGS-SNP-biallelic-pruned-10Mb-1
```

## Future

Ru-run if (`rerun_plink`) is fun for the future:

```
plink --rerun run1.log --maf 0.1
plink --rerun run1.log --maf 0.1 --out run2
```

Run PLINK from script, not now (`run_plink_from_script`):

```
plink --script myscript1.txt
```

> where the file myscript1.txt is a plain text file containing

```
--ped ..\data\version1\50K\allsamples.ped 
--map ..\data\allmapfiles\finalversion\autosomal.map 
--out ..\results\working\sample-missingness-v1.22
--from rs66537222
--to rs8837323 
--geno 0.25
--maf 0.02
--missing
```
